src/bootstrap/compile.rs

```diff
@@ -1064,15 +1064,9 @@ pub fn run_cargo(builder: &Builder,
 
             let filename = Path::new(&*filename);
 
-            // If this was an output file in the "host dir" we don't actually
-            // worry about it, it's not relevant for us.
-            if filename.starts_with(&host_root_dir) {
-                continue;
-            }
-
             // If this was output in the `deps` dir then this is a precise file
             // name (hash included) so we start tracking it.
-            if filename.starts_with(&target_deps_dir) {
+            if filename.starts_with(&host_root_dir) || filename.starts_with(&target_deps_dir) {
```

Comment:
Can you expand a bit on what's going on here? This feels like it shouldn't be necessary...

Reply:
The PR description tries to explain this. It's needed for the compiler to be able to find the (host-only) proc macro dependencies of e.g. `librustc`, from across a Cargo build (i.e. from codegen backends).

Reply:
Hm ok, I ask because I don't think that this is right. This I think is definitely not what we want when cross-compiling because it's mixing architectures of libraries in one directory. For non-cross-compiling situations this'll copy too much for things like build scripts and/or intermediate proc-macro dependencies.

Could we perhaps whitelist an explicit copy of serde_derive if necessary? Or something like that?

Reply:
It would be nice if cargo would tell us if artifacts are proc macros, but I think a whitelist will suffice here

Reply:
Mixing architectures isn't a problem for `rustc`, and a whitelist feels hackier IMO.

I still like the idea of making the same `-Ldependency=` distinction Cargo makes, in the sysroot, as well (from https://github.com/rust-lang/rust/pull/56447#issuecomment-443619921).

In fact, we can even copy the host deps to `rustlib/$host/deps` instead of `rustlib/$target/deps`, and have the compiler look in the right place.

<hr/>

I should also mention I was able to work around this by adding a `serde_derive` dependency to `librustc_codegen_llvm`, but I feel like this also doesn't scale (and it builds `serde_derive` more than once).

<hr/>

I guess I'd accept the whitelisting if someone else implemented it, just to unblock @Zoxc.